src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs

```diff
@@ -9,50 +9,50 @@
 using System.Net.Http.Headers;
 using System.Text;
 using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
 using System.Threading;
 using System.Threading.Tasks;
 
 namespace System.Net.Http.Json
 {
     public sealed partial class JsonContent : HttpContent
     {
-        private readonly JsonSerializerOptions? _jsonSerializerOptions;
-        public Type ObjectType { get; }
+        private readonly JsonTypeInfo _typeInfo;
+        public Type ObjectType => _typeInfo.Type;
         public object? Value { get; }
 
-        [RequiresUnreferencedCode(HttpContentJsonExtensions.SerializationUnreferencedCodeMessage)]
-        [RequiresDynamicCode(HttpContentJsonExtensions.SerializationDynamicCodeMessage)]
         private JsonContent(
             object? inputValue,
-            Type inputType,
-            MediaTypeHeaderValue? mediaType,
-            JsonSerializerOptions? options)
+            JsonTypeInfo jsonTypeInfo,
+            MediaTypeHeaderValue? mediaType)
         {
-            if (inputType is null)
-            {
-                throw new ArgumentNullException(nameof(inputType));
-            }
+            ThrowHelper.ThrowIfNull(jsonTypeInfo);
 
-            if (inputValue != null && !inputType.IsAssignableFrom(inputValue.GetType()))
+            if (inputValue != null && !jsonTypeInfo.Type.IsAssignableFrom(inputValue.GetType()))
```

Comment:
This check is going to be redundant for the case of the strongly type `JsonTypeInfo<T>` APIs and `JsonSerializerOptions` APIs. I'd replaced this with a `Debug.Assert()` and only include validation in the relevant weakly typed APIs that call into the constructor.

Reply:
Please also include a test that validates that this exception is being thrown where possible.

Reply:
> This check is going to be redundant for the case of the strongly type `JsonTypeInfo<T>` APIs and `JsonSerializerOptions` APIs. I'd replaced this with a `Debug.Assert()` and only include validation in the relevant weakly typed APIs that call into the constructor.

Done